sunpy/net/helioviewer.py

```diff
@@ -145,42 +151,86 @@ def download_jp2(self, date, sourceid, directory=None, overwrite=False, **kwargs
         ----------
         date : `datetime.datetime`, string
             A string or datetime object for the desired date of the image
-        sourceid : int
-            It is a unique number that can be used to refer a particular image
-            datasource in the API reqeust.
         directory : string
             (Optional) Directory to download JPEG 2000 image to.
-        json : bool
-            (Optional) Returns a JSON object if set to True
+        observatory : string
+            (Optional) Observatory name
+        instrument : string
+            (Optional) instrument name
+        detector : string
+            (Optional) detector name
+        measurement : string
+            (Optional) measurement name
+        sourceId : int
+            (Optional) data source id
+        jpip : bool
+            (Optional) Returns a JPIP URI if set to True
 
         Returns
         -------
         out : string
-            Returns a filepath to the downloaded JPEG 2000 image.
+            Returns a filepath to the downloaded JPEG 2000 image or a URL if
+            the "jpip" parameter is set to True.
 
         Examples
         --------
         >>> import sunpy.map
         >>> from sunpy.net import helioviewer
         >>> hv = helioviewer.HelioviewerClient()  # doctest: +REMOTE_DATA
-        >>> source_id = hv.get_source_id()  # doctest: +REMOTE_DATA
-        >>> hv.display_source_id()  # doctest: +SKIP
-        >>> file = hv.download_jp2('2012/07/03 14:30:00', sourceid = source_id[('SDO', 'HMI', 'continuum')])   # doctest: +REMOTE_DATA
-        >>> aia = sunpy.map.Map(file)   # doctest: +REMOTE_DATA
+        >>> filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', instrument='HMI', detector='None', measurement='continuum')   # doctest: +REMOTE_DATA
+        >>> aia = sunpy.map.Map(filepath)   # doctest: +REMOTE_DATA
         >>> aia.peek()   # doctest: +SKIP
+        >>> data_sources = hv.get_source_dict()  # doctest: +REMOTE_DATA
+        >>> file = hv.download_jp2('2012/07/03 14:30:00', sourceId=data_sources[('SOHO' 'LASCO' 'C2' 'white-light')]   # doctest: +REMOTE_DATA
+        """
+        if('observatory' in kwargs):
+            observatory = kwargs['observatory'] if(kwargs['observatory']) else 'None' 
+            instrument = kwargs['instrument'] if(kwargs['instrument']) else 'None'
+            detector = kwargs['detector'] if(kwargs['detector']) else 'None'
+            measurement = kwargs['measurement'] if(kwargs['measurement']) else 'None'
+            key = (observatory, instrument, detector, measurement)
+            filepath = self.download_jp2_(date, self._source_dict[key], directory, overwrite)
+        elif('sourceId' in kwargs):
+            key = kwargs['sourceId']
+            filepath = self.download_jp2_(date, key, directory, overwrite)
+        
+        return filepath
+
+    def download_jp2_(self, date, sourceid, directory=None, overwrite=False, **kwargs):
```

Comment:
Why was this created?

Reply:
I created this because of the user tries to download the image using observatory, instruments etc then the key for the dictionary will be different and if the user wants to use source ID then the key will be different